core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -984,11 +984,11 @@ private List<ManifestFile> newDataFilesAsManifests() {
                 writer.close();
               }
               this.cachedNewDataManifests.addAll(writer.toManifestFiles());
-              this.hasNewDataFiles = false;
```

Comment:
I don't think it caused issues but it would be more correct to set `hasNewDataFiles` to `false` after processing all new data files, similarly to what we do for delete files. Otherwise, it is confusing.

Reply:
Yeah agreed, this change is easier to read. Also agree that setting it to false prematurely shouldn't have caused issues, looking at how `hasNewDataFiles` is used